3rdParty/fuerte/src/AsioSockets.h

```diff
@@ -77,22 +91,41 @@ struct Socket<SocketType::Tcp> {
   Socket(EventLoopService&, asio_ns::io_context& ctx)
       : resolver(ctx), socket(ctx), timer(ctx) {}
 
-  ~Socket() { this->cancel(); }
+  ~Socket() { 
+    try {
+      this->cancel(); 
+    } catch (std::exception const& ex) {
+      FUERTE_LOG_ERROR << "caught exception during tcp socket shutdown: " << ex.what();
+    }
+  }
 
   template <typename F>
   void connect(detail::ConnectionConfiguration const& config, F&& done) {
-    resolveConnect(config, resolver, socket, std::forward<F>(done));
+    resolveConnect(config, resolver, socket, [this, done = std::forward<F>(done)](asio_ns::error_code ec) mutable {
+      FUERTE_LOG_DEBUG << "executing tcp connect callback, ec: " << ec.message() << ", canceled: " << this->canceled;
+      if (canceled) {
+        // cancel() was already called on this socket
+        ec = asio_ns::error::operation_aborted;
```

Comment:
I would like to add an assertion that `is_open() == false` here.

Reply:
Isn't it possible that both the connect succeed and the timeout is triggered, so both callbacks are scheduled. In this case, it is possible that the timeout is scheduled first and closes the socket and your assertion here would fail?

Reply:
No, this is nonense, please ignore. THis is the "cancelled == true" case, in this case the socket must be closed.